### DISCUSSION TITLE
RawInline openxml merge with adjacent run (`<w:r>...</w:r>`) for better tabstop support?

### DISCUSSION THREAD

**Discussion by jblachly** [jgm/pandoc#9985] [Q&A] [answered]
# Background

I am converting HTML to Docx and want to support tabstops correctly

# Problem statement:

I'd like to have both left-aligned text, as well as right-aligned text using a tab stop. This looks like this:

![Screenshot 2024-07-14 at 10 22 48 PM](https://github.com/user-attachments/assets/3b2b67db-0c10-44d7-b208-497a8177d35f)

# Imperfect solution

I have figured out how to do this by modifying the reference.docx and inserting the appropriate tab stop; then building a Lua filter that looks for `<span>\tab</span>` and replaces that node with `pandoc.RawInline('openxml', '<w:r><w:tab /></w:r>')`. This obviously inserts the XML you see into the document between runs on either side of the span.

This works well unless the left text is too long, in which case Word's render engine exhibits a strange artifact:
![Screenshot 2024-07-14 at 10 27 39 PM](https://github.com/user-attachments/assets/4cde2237-f31b-493c-a1d3-b0bdef8050d0)

To confirm this is not how Word behaves by default, I re-created this in a brand new Word doc, and you see the correct behavior (viz., the right-sided text immediately flows to the next line):

![Screenshot 2024-07-14 at 10 31 17 PM](https://github.com/user-attachments/assets/4b8dd796-2ddc-4a2a-9d50-087a5831566b)

# Dissecting the OOXML

The pandoc-emitted XML looks like this:
```xml
        <w:p>
            <w:pPr>
                <w:pStyle w:val="BodyText" />
            </w:pPr>
            <w:r>
                <w:t xml:space="preserve">Some left-sided text</w:t>
            </w:r>
            <w:r>
                <w:tab />
            </w:r>
            <w:r>
                <w:t xml:space="preserve">2003 — 2007</w:t>
            </w:r>
        </w:p>
```

whereas the default MS Word Docx XML looks like this:
(the `w:tabs` does not appear in pandoc output because it is applied via style)

```xml
        <w:p w14:paraId="662A0B1C" w14:textId="359A6CE4" w:rsidR="00CE3EA9"
            w:rsidRDefault="0045750A" w:rsidP="0045750A">
            <w:pPr>
                <w:tabs>
                    <w:tab w:val="right" w:pos="9360" />
                </w:tabs>
            </w:pPr>
            <w:r>
                <w:t xml:space="preserve">This is left aligned text</w:t>
            </w:r>
            <w:r>
                <w:tab />
                <w:t>This is right aligned text</w:t>
            </w:r>
        </w:p>
```

**!!! note that the `<w:tab />` is contained within the same run as the R-sided text !!!** -- This is the crucial difference in the Word render engine behavior.

# HELP

What I need at this point is a way to insert a raw OpenXML tag (`w:tab`) _inside an adjacent `<w:r>` run_.  I have tried several (probably ill-informed) approaches to no avail.

I'm throwing in the towel for now, but I wanted to propose ask about something I haven't tried:

Could I somehow wrap everything downstream of the `\tab` span (which is essentially a bunch of `Str` and `Space` nodes in my case) up with the tab into a new `Span`? Would this cause the [OpenXML writer](src/Text/Pandoc/Writers/Docx/OpenXML.hs) to enclose them all together in a single `<w:r>...</w:r>` pair?

**Comment by jgm** [comment]
How did you generate this with pandoc? There's only one place in the code where `w:tab` is emitted, and that's to separate the number and the section name in a heading.

**Reply by jblachly** [reply]
Sorry if I was not clear (see "Imperfect solution").

Explicitly, with the following luascript:

```lua
--[[
	Transform inline
		<span>\tab</span>
	into docx tab

  NOTE:
  No extra spaces inside the span!

  I documented at some point that pandoc's AST for HTML apparently never includes RawBlock (see pagebreak.lua)

  REFERENCES:
   - https://github.com/blachlylab/wodehouse/issues/427
   - https://github.com/jgm/pandoc/discussions/9985
]]
local ooxml = {
  --tab = '<w:tab />' doesn't work, must include wrapping run, but see issue and discussion in REFERENCES above
  tab = '<w:r><w:tab /></w:r>'
}

return {
  {Span = function (el)
    if el.content[1].text == "\\tab" then
      return pandoc.RawInline('openxml', ooxml.tab)
      --return pandoc.List({
      --  pandoc.RawInline('plain', '<w:tab />')
      --})
    else
      return el
    end
  end,
  }
}
```

operating on a document like:
```html
<p>Hello</p>
<p>
this is left text<span>\tab</span>this is right
</p>
<p>
Left text 2 <span>\tab Right text 2</span>
</p>
```

**Reply by jgm** [reply, marked as answer]
Got it, thanks.  I don't see any good solution short of parsing and rewriting the XML in the docx container.  In theory one could rewrite the openxml writer so that it automatically consolidates w:r elements if they don't contain style information.  But it wouldn't be a trivial rewrite, given how things are currently done.

**Reply by jblachly** [reply]
> rewrite the openxml writer so that it automatically consolidates w:r elements if they don't contain style information

Thank you. I would tackle this if I had the bandwidth to learn Haskell. Given that it's "non-trivial", I'd better steer clear.
